### ISSUE TITLE
Wrong normalization in FFT methods

### ISSUE THREAD

**Issue by unalmis** [google/jax#22466]
### Description

Summary: The discrete cosine transformation libraries normalize the transforms incorrectly.

I am aware that JAX's documentation states that only `norm="ortho"` is supported, however, this still requires the user to explicitly pass `norm=ortho`, which the documentation does not state. To be clear, jax gives incorrect results unless `norm="ortho"` is passed explicitly. Jax also doesn't seem to implement backward dct and forward idct correctly. If this is not supported yet, then JAX should fail early. The effect of these bugs is that interpolation with fft's without specifying additional flags gives incorrect results.

Minimal reproductions.

Bug 1.
```python
import numpy as np
from scipy.fftpack import dct as sdct
from scipy.fftpack import idct as sidct
from jax.scipy.fft import dct as jdct
from jax.scipy.fft import idct as jidct
import matplotlib.pyplot as plt

m = np.linspace(0, 2 * np.pi, 100)
n = np.linspace(0, 2 * np.pi, 5 * m.size)
f = m
normalize = np.sqrt(n.size / m.size)
sf = sidct(sdct(f, norm="ortho"), n=n.size, norm="ortho") * normalize
jf = jidct(jdct(f, norm="ortho"), n=n.size, norm="ortho") * normalize
# Doesn't implement ortho correctly unless specifed. Documentation says only ortho is supported. Bug 1.
jf_bug = jidct(jdct(f), n=n.size) * normalize

plt.scatter(m, f, label="truth", color="black")
plt.plot(n, sf, linestyle="-", linewidth=4, label="ortho scipy")
plt.plot(n, jf, linestyle="--", linewidth=2.5, label="ortho jax")
plt.plot(n, jf_bug, linestyle="--", linewidth=2.5, label="ortho jax bug")
plt.legend()
```
![out](https://github.com/user-attachments/assets/919fc36f-613a-4ece-8924-315cc5804d96)

Ok, so now the user wonders whether the documentation is out of date, and tries to find out what is the normalization that JAX implements by default. They may conclude that default is `backward` on `dct` and `forward` on `idct`. Scipy's default is `backward` for both. This departure from `scipy` seems confusing given that this methods are in `jax.scipy.fft`.

More confusing, is that when we specify explicitly to do `backward` on `dct` and `forward` on `idct` for `jax`, the results change, so it seems JAX does not implement the options that is does by default correctly when they are specified explicitly.

Bug 2.
```python
import numpy as np
from scipy.fftpack import dct as sdct
from scipy.fftpack import idct as sidct
from jax.scipy.fft import dct as jdct
from jax.scipy.fft import idct as jidct
import matplotlib.pyplot as plt

m = np.linspace(0, 2 * np.pi, 100)
n = np.linspace(0, 2 * np.pi, 5 * m.size)
f = m
normalize = n.size / m.size
sf = sidct(sdct(f, norm="backward"), n=n.size, norm="forward") * normalize
jf_bug = jidct(jdct(f), n=n.size) * normalize  # matches sf, so must be doing backward then forward
# But this does not match the above 2.
jf_bug_2 = jidct(jdct(f, norm="backward"), n=n.size, norm="forward") * normalize

plt.scatter(m, f, label="truth", color="black")
plt.plot(n, sf, linestyle='-', linewidth=4, label="scipy backward forward")
# This plot matches the truth, but only because we multiplied by "normalize",
# if it was doing ortho then it should only match if we multiply by sqrt(normalize),
# hence the bug
plt.plot(n, jf_bug, linestyle='--', linewidth=2.5, label="ortho jax bug")
# If jax doesn't support backward and forward, then this should fail early with an error.
plt.plot(n, jf_bug_2, linestyle='--', linewidth=2.5, label="backward forward jax bug 2")
plt.legend()
```
![out2](https://github.com/user-attachments/assets/7e87ae64-a7e9-49f0-a376-cdcfd352101b).

Please let me know if I am missing something.

### System info (python version, jaxlib version, accelerator, etc.)

```
An NVIDIA GPU may be present on this machine, but a CUDA-enabled jaxlib is not installed. Falling back to cpu.
jax:    0.4.28
jaxlib: 0.4.28
numpy:  1.26.4
python: 3.10.14 | packaged by conda-forge | (main, Mar 20 2024, 12:45:18) [GCC 12.3.0]
jax.devices (1 total, 1 local): [CpuDevice(id=0)]
process_count: 1
platform: uname_result(system='Linux', node='fedora', release='6.9.8-200.fc40.x86_64', version='#1 SMP PREEMPT_DYNAMIC Fri Jul  5 16:20:11 UTC 2024', machine='x86_64')
```

**Comment by unalmis** [comment]
For the second example, in the newer `scipy.fft` libraries, the equivalent of
```python
# with scipy.fftpack
normalize = n.size / m.size
sf = sidct(sdct(f, norm="backward"), n=n.size, norm="forward") * normalize
```
becomes
```python
# with scipy.fft
normalize = n.size / m.size
sf = sidct(sdct(f, norm="backward"), n=n.size, norm="backward") * normalize
```

Still, replacing
```python
jf_bug_2 = jidct(jdct(f, norm="backward"), n=n.size, norm="forward") * normalize
```
with 
```python
jf_bug_2 = jidct(jdct(f, norm="backward"), n=n.size, norm="backward") * normalize
```
does not change any issues raised in the original post.

**Comment by mattjj** [comment]
Thanks for reporting this so clearly!

**Comment by unalmis** [comment]
It's unrelated to this issue, but on the topic of `ortho="norm"`, please also be aware of https://github.com/scipy/scipy/issues/21198. Unlike, scipy, JAX doesn't provide the `orthogonalize=False` option. This makes me skeptical of using JAX's DCT (even for type 2 DCT). The two linked issues shown there only discuss the effect on the type 1 DCT, but I think the same issues should occur to types 1,2,3.

Ideally, I'd like set `ortho="backward"` to avoid this issue, but given that JAX only implements `ortho="norm"` I think JAX should implement an option to set `orthogonalize=False`.

Should I make a new issue from this comment for this?

**Comment by mattjj** [comment]
> Should this comment be referenced in a new issue?

Yes, new issue please!

**Comment by Saanidhyavats** [comment]
Hello everyone. I have recently started using JAX and interested to contribute in this community. @mattjj can I start working on this issue? Is there anything you would suggest me before starting on it?

**Comment by jakevdp** [comment]
Thanks for the report. I've added validation that raises an error for unsupported `norm` values in #22567.

As for the incorrect results with `norm=None`: this is strange because we have a suite of tests against scipy's implementation for the default `norm=None`; for example https://github.com/google/jax/blob/db05734041a6b6c26440b087a22c036825665dc2/tests/scipy_fft_test.py#L48-L62

There must be some relevant case in that test parameterization that is not covered.

**Comment by jakevdp** [comment]
Ah, it looks like our implementation does match that of `scipy.fft`, although it does not match `scipy.fftpack`:
```python
import numpy as np
from scipy.fft import dct as sdct
from scipy.fft import idct as sidct
from jax.scipy.fft import dct as jdct
from jax.scipy.fft import idct as jidct
import matplotlib.pyplot as plt

m = np.linspace(0, 2 * np.pi, 100)
n = np.linspace(0, 2 * np.pi, 5 * m.size)
f = m
normalize = np.sqrt(n.size / m.size)

jf_none = jidct(jdct(f), n=n.size) * normalize
sf_none = sidct(sdct(f), n=n.size) * normalize

plt.plot(n, sf_none, linestyle="-", linewidth=4, label="ortho scipy")
plt.plot(n, jf_none, linestyle="--", linewidth=2.5, label="ortho jax")
plt.legend()
```
![download](https://github.com/user-attachments/assets/8f073f27-85c6-44f9-b34e-e59dfa736088)

I believe that `scipy.fftpack` is deprecated, so you should be using the `scipy.fft` versions. With that in mind, I think the only bug here involves validation of unsupported `norm` arguments, which is fixed in #22567.

If you can find any discrepancies between the outputs of `jax.scipy.fft` functions and `scipy.fft` functions, please let us know!

**Comment by jakevdp** [comment]
#22571 implements and tests `norm='backward'` for all `dct` functions, and ensures that its behavior matches the corresponding implementations in `scipy.fft`.

**Comment by unalmis** [comment]
@jakevdp Thanks for fixing this! It looks like #22567 and #22571 should resolve the issue. Perhaps it's useful to document: #22567 alone would not solve the issue:

After the first PR,
1. The documentation states that JAX only supports `norm="ortho"`.
2. It's good that JAX now fails if an unsupported argument is given. However, note that JAX will now fail for all options `norm not in ["ortho", None]`.
3. 1 and 2 both imply that by default, `norm="ortho"` will be performed. This differs from `scipy.fft`, which states that, by default, `norm="backward"`. In that case matching plots you showed in your previous comment would indicate a bug, because the two plots should differ. We should instead expect the following two plots to match.
```python
import numpy as np
from scipy.fft import dct as sdct
from scipy.fft import idct as sidct
from jax.scipy.fft import dct as jdct
from jax.scipy.fft import idct as jidct
import matplotlib.pyplot as plt

m = np.linspace(0, 2 * np.pi, 100)
n = np.linspace(0, 2 * np.pi, 5 * m.size)
f = m
# scipy has backward is default
sf = sidct(sdct(f), n=n.size) * (n.size / m.size)
# jax uses ortho by default
jf = jidct(jdct(f), n=n.size) * np.sqrt(n.size / m.size)

plt.scatter(m, f, label="truth", color="black")
plt.plot(n, sf, linestyle="-", linewidth=4, label="scipy")
plt.plot(n, jf, linestyle="--", linewidth=2.5, label="jax")
plt.legend()
```
![image](https://github.com/user-attachments/assets/d0949014-b55e-4ca9-acec-d215aeff502d)


The second PR updates the documentation to state `norm="backward"` is supported and used by default. In that case, the correct usage becomes
```python
sf = sidct(sdct(f), n=n.size) * (n.size / m.size)
jf = jidct(jdct(f), n=n.size) * (n.size / m.size)
```
which produces matching plots as expected.

**Comment by jakevdp** [comment]
I think https://github.com/google/jax/pull/22571 addresses all your comments – please let me know if that is not the case.

**Comment by unalmis** [comment]
Yes it does, thanks.
